### DISCUSSION TITLE
Clarify the differences between the supported control modes

### DISCUSSION THREAD

**Discussion by diegoferigo** [bulletphysics/bullet3#3649] [General]
Reading the documentation, it is not entirely clear to me what are the differences of the joint control modes supported in pybullet.

I try to recap them here, please @erwincoumans feel free to jump in to add more details and address holes in my understanding.

Let's assume that in all cases (excluding torque), the reference error is the following:

```
error = kp *  (pos_des - pos) + kd * (vel_des - vel)
```

- **`VELOCITY_CONTROL`**:  This is a pure proportional control for velocity that by default has `kp=0` and `kd=1`. Gains should be in the range `[0, 1]`.
- **`POSITION_CONTROL`** (alias for `CONTROL_MODE_POSITION_VELOCITY_PD`):  Similar to `VELOCITY_CONTROL`. By default it has `kp=1` and `kd=0`. It allows to optionally specify both the target position and target velocity.
- **`TORQUE_CONTROL`**: It should be enabled by disabling the default motorized joints passing a `force=0`. Later calls using this mode will send directly the torque to the joint.
- **`PD_CONTROL`**: Similar to `POSITION` control but it allows specifying not normalized gains.

Questions:
- Is `kp` ignored in `VELOCITY_CONTROL`?
- What is the logic behind the normalized gains in `VELOCITY_CONTROL` and `POSITION_CONTROL`?
- Is `kd=0` by default in `POSITION_CONTROL`?
- Is correct that in `POSITION_CONTROL` if `kd=0` and a velocity reference is specified, it has no effect?
- Is the description of `PD_CONTROL` correct? There's not much information about it.
- Should `PD_CONTROL` with `kp=0` and `kd=1` behave as `VELOCITY_CONTROL`?

I think that a big difference of `PD_CONTROL` with respect to `POSITION` and `VELOCITY` is how the error is treated. I suspect that in the former case a joint force is generated, in the latter ones instead the error is minimized as constraint. Is this correct?

Related issues https://github.com/bulletphysics/bullet3/issues/2152 https://github.com/bulletphysics/bullet3/issues/1026

**Comment by Steven89Liu** [comment]
i also have these questions, thanks in advance.

**Comment by raunaqbhirangi** [comment]
I think it's evident from the switch case [here](https://github.com/bulletphysics/bullet3/blob/master/examples/pybullet/pybullet.c#L3436) that PD_CONTROL and POSITION_CONTROL should have identical behavior.

I had an additional query. When using PD_CONTROL, the getJointState function seems to be returning zeros for the appliedJointMotorTorque return value. Is this intended behavior? I had imagined that this function would work for PD_CONTROL in the same way that it does for POSITION_CONTROL and VELOCITY_CONTROL. Not sure if this should be a separate issue.

**Comment by raffaello-camoriano** [comment]
Same questions here.

**Comment by zhangy76** [comment]
Same questions here.

**Comment by mahdinobar** [comment]
same question. who is the maintainer of this repo?

**Comment by salehrayan** [comment]
Don't know why these important questions are not addressed yet. maybe based on [this](https://github.com/bulletphysics/bullet3/blob/e9c461b0ace140d5c73972760781d94b7b5eee53/examples/pybullet/pybullet.c#L2368-L2376) the default kp and kd is 0.1 and 1 in position control?

**Comment by cc299792458** [comment]
same question.
